File: src/core/rng.ts
/**
 * Deterministic, serializable pseudo-random number generator.
 *
 * Every random decision in the rules layer flows through this so that a battle
 * can be replayed exactly from a seed. That matters for netplay, for replays,
 * and for reproducing balance-test failures. The generator is a 128-bit
 * xoshiro-style state advanced with a 32-bit mixing step; it is not
 * cryptographically secure and is not intended to be.
 */

export interface RngState {
  a: number;
  b: number;
  c: number;
  d: number;
}

const UINT32 = 0x100000000;

function rotl(x: number, k: number): number {
  return ((x << k) | (x >>> (32 - k))) >>> 0;
}

/** Build a generator state from an arbitrary string or numeric seed. */
export function seedRng(seed: string | number): RngState {
  let h = 0x9e3779b9 >>> 0;
  const text = typeof seed === 'number' ? String(seed) : seed;
  for (let i = 0; i < text.length; i++) {
    h = Math.imul(h ^ text.charCodeAt(i), 0x85ebca6b) >>> 0;
    h = (h ^ (h >>> 13)) >>> 0;
  }
  const next = () => {
    h = (h + 0x6d2b79f5) >>> 0;
    let t = h;
    t = Math.imul(t ^ (t >>> 15), t | 1) >>> 0;
    t = (t ^ (t + Math.imul(t ^ (t >>> 7), t | 61))) >>> 0;
    return (t ^ (t >>> 14)) >>> 0;
  };
  const state: RngState = { a: next(), b: next(), c: next(), d: next() };
  // Guard against the all-zero state, which is a fixed point.
  if (!state.a && !state.b && !state.c && !state.d) state.a = 0x1a2b3c4d;
  return state;
}

/** Copy a state so callers can branch a timeline without mutating the original. */
export function cloneRng(state: RngState): RngState {
  return { a: state.a, b: state.b, c: state.c, d: state.d };
}

/** Advance the state and return a uint32. Mutates `state` in place. */
export function nextUint32(state: RngState): number {
  const t = (state.b << 9) >>> 0;
  state.c = (state.c ^ state.a) >>> 0;
  state.d = (state.d ^ state.b) >>> 0;
  state.b = (state.b ^ state.c) >>> 0;
  state.a = (state.a ^ state.d) >>> 0;
  state.c = (state.c ^ t) >>> 0;
  state.d = rotl(state.d, 11);
  const result = (rotl(Math.imul(state.b, 5) >>> 0, 7) * 9) >>> 0;
  return result;
}

/** Uniform float in [0, 1). */
export function nextFloat(state: RngState): number {
  return nextUint32(state) / UINT32;
}

/** Uniform integer in [min, max] inclusive. */
export function nextInt(state: RngState, min: number, max: number): number {
  if (max < min) throw new RangeError(`nextInt: max (${max}) < min (${min})`);
  const span = max - min + 1;
  return min + Math.floor(nextFloat(state) * span);
}

/** True with the given probability (0..1). */
export function chance(state: RngState, probability: number): boolean {
  if (probability <= 0) return false;
  if (probability >= 1) return true;
  return nextFloat(state) < probability;
}

/** Pick one element. Returns undefined for an empty list. */
export function pick<T>(state: RngState, items: readonly T[]): T | undefined {
  if (items.length === 0) return undefined;
  return items[nextInt(state, 0, items.length - 1)];
}

/** Fisher-Yates into a new array; the input is left untouched. */
export function shuffle<T>(state: RngState, items: readonly T[]): T[] {
  const out = items.slice();
  for (let i = out.length - 1; i > 0; i--) {
    const j = nextInt(state, 0, i);
    const a = out[i] as T;
    const b = out[j] as T;
    out[i] = b;
    out[j] = a;
  }
  return out;
}

/**
 * Weighted pick. Weights must be non-negative; entries with weight 0 are never
 * chosen. Returns undefined if every weight is 0 or the list is empty.
 */
export function weightedPick<T>(
  state: RngState,
  items: readonly T[],
  weightOf: (item: T) => number,
): T | undefined {
  let total = 0;
  for (const item of items) total += Math.max(0, weightOf(item));
  if (total <= 0) return undefined;
  let roll = nextFloat(state) * total;
  for (const item of items) {
    roll -= Math.max(0, weightOf(item));
    if (roll < 0) return item;
  }
  return items[items.length - 1];
}

/** Symmetric variance multiplier, e.g. spread 0.1 gives a factor in [0.9, 1.1). */
export function variance(state: RngState, spread: number): number {
  return 1 + (nextFloat(state) * 2 - 1) * spread;
}

File: unity/Aetherlight.Core/Core/Rng.cs
using System;
using System.Collections.Generic;

namespace Aetherlight.Core
{
    /// <summary>
    /// Deterministic, serializable pseudo-random number generator.
    ///
    /// This is a bit-exact port of the TypeScript original. That exactness is
    /// the whole point: seeds, replays, and balance-test failures are only
    /// portable between the two implementations if the generators agree on
    /// every bit. <c>RngParityTests</c> pins that against a fixture captured
    /// from the TypeScript side.
    ///
    /// Porting notes, since JavaScript's integer semantics do not survive a
    /// naive translation:
    ///   * <c>Math.imul(a, b)</c> is a 32-bit multiply that wraps. C# <c>uint</c>
    ///     multiplication inside <c>unchecked</c> produces the same bit pattern.
    ///   * <c>x >>> 0</c> coerces to uint32; using <c>uint</c> throughout makes
    ///     it implicit.
    ///   * Addition and XOR agree between the two languages once both sides are
    ///     mod-2^32, because two's complement addition is sign-agnostic.
    /// </summary>
    public struct RngState : IEquatable<RngState>
    {
        public uint A;
        public uint B;
        public uint C;
        public uint D;

        public RngState(uint a, uint b, uint c, uint d)
        {
            A = a;
            B = b;
            C = c;
            D = d;
        }

        public bool Equals(RngState other) => A == other.A && B == other.B && C == other.C && D == other.D;

        public override bool Equals(object? obj) => obj is RngState other && Equals(other);

        public override int GetHashCode() => unchecked((int)(A ^ B ^ C ^ D));

        public override string ToString() => $"RngState({A}, {B}, {C}, {D})";
    }

    public static class Rng
    {
        private const double Uint32Range = 4294967296.0;

        private static uint Rotl(uint x, int k) => unchecked((x << k) | (x >> (32 - k)));

        /// <summary>Build a generator state from a string or numeric seed.</summary>
        public static RngState Seed(string seed)
        {
            if (seed == null) throw new ArgumentNullException(nameof(seed));

            uint h = 0x9e3779b9u;
            unchecked
            {
                for (int i = 0; i < seed.Length; i++)
                {
                    h = (h ^ seed[i]) * 0x85ebca6bu;
                    h = h ^ (h >> 13);
                }
            }

            uint Next()
            {
                unchecked
                {
                    h += 0x6d2b79f5u;
                    uint t = h;
                    t = (t ^ (t >> 15)) * (t | 1u);
                    t = t ^ (t + (t ^ (t >> 7)) * (t | 61u));
                    return t ^ (t >> 14);
                }
            }

            var state = new RngState(Next(), Next(), Next(), Next());
            // The all-zero state is a fixed point; nudge it off.
            if (state.A == 0 && state.B == 0 && state.C == 0 && state.D == 0) state.A = 0x1a2b3c4du;
            return state;
        }

        public static RngState Seed(long seed) => Seed(seed.ToString(System.Globalization.CultureInfo.InvariantCulture));

        /// <summary>Advance the state and return a uint32.</summary>
        public static uint NextUInt32(ref RngState state)
        {
            unchecked
            {
                uint t = state.B << 9;
                state.C ^= state.A;
                state.D ^= state.B;
                state.B ^= state.C;
                state.A ^= state.D;
                state.C ^= t;
                state.D = Rotl(state.D, 11);
                return Rotl(state.B * 5u, 7) * 9u;
            }
        }

        /// <summary>Uniform double in [0, 1).</summary>
        public static double NextDouble(ref RngState state) => NextUInt32(ref state) / Uint32Range;

        /// <summary>Uniform integer in [min, max], inclusive.</summary>
        public static int NextInt(ref RngState state, int min, int max)
        {
            if (max < min) throw new ArgumentOutOfRangeException(nameof(max), $"NextInt: max ({max}) < min ({min})");
            long span = (long)max - min + 1;
            return min + (int)Math.Floor(NextDouble(ref state) * span);
        }

        /// <summary>True with the given probability (0..1).</summary>
        public static bool Chance(ref RngState state, double probability)
        {
            if (probability <= 0) return false;
            if (probability >= 1) return true;
            return NextDouble(ref state) < probability;
        }

        /// <summary>Pick one element, or default for an empty list.</summary>
        public static T? Pick<T>(ref RngState state, IReadOnlyList<T> items) where T : class
        {
            if (items == null || items.Count == 0) return null;
            return items[NextInt(ref state, 0, items.Count - 1)];
        }

        /// <summary>Fisher-Yates into a new list; the input is left untouched.</summary>
        public static List<T> Shuffle<T>(ref RngState state, IReadOnlyList<T> items)
        {
            var outList = new List<T>(items);
            for (int i = outList.Count - 1; i > 0; i--)
            {
                int j = NextInt(ref state, 0, i);
                (outList[i], outList[j]) = (outList[j], outList[i]);
            }
            return outList;
        }

        /// <summary>
        /// Weighted pick. Weights must be non-negative; zero-weight entries are
        /// never chosen. Returns false if every weight is zero or the list is empty.
        /// </summary>
        public static bool TryWeightedPick<T>(ref RngState state, IReadOnlyList<T> items, Func<T, double> weightOf, out T result)
        {
            result = default!;
            if (items == null || items.Count == 0) return false;

            double total = 0;
            foreach (var item in items) total += Math.Max(0, weightOf(item));
            if (total <= 0) return false;

            double roll = NextDouble(ref state) * total;
            foreach (var item in items)
            {
                roll -= Math.Max(0, weightOf(item));
                if (roll < 0)
                {
                    result = item;
                    return true;
                }
            }
            result = items[items.Count - 1];
            return true;
        }

        /// <summary>Symmetric variance multiplier: spread 0.1 gives [0.9, 1.1).</summary>
        public static double Variance(ref RngState state, double spread) => 1 + (NextDouble(ref state) * 2 - 1) * spread;
    }
}
